src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVMKernels.hpp

```diff
@@ -120,69 +130,720 @@ struct FluxKernel
+
+  /// Views on dof numbers
+  ElementViewConst< arrayView1d< globalIndex const > > const m_dofNumber;
+
+  /// Views on permeability
+  ElementViewConst< arrayView3d< real64 const > > m_permeability;
+  ElementViewConst< arrayView3d< real64 const > > m_dPerm_dPres;
+
+  /// Views on ghost rank numbers and gravity coefficients
+  ElementViewConst< arrayView1d< integer const > > const m_ghostRank;
+  ElementViewConst< arrayView1d< real64 const > > const m_gravCoef;
+
+  // Primary and secondary variables
+
+  /// Views on pressure
+  ElementViewConst< arrayView1d< real64 const > > const m_pres;
+  ElementViewConst< arrayView1d< real64 const > > const m_dPres;
+
+  /// Views on derivatives of phase volume fractions and comp fractions
+  ElementViewConst< arrayView3d< real64 const, compflow::USD_COMP_DC > > const m_dCompFrac_dCompDens;
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_PHASE > > const m_dPhaseVolFrac_dPres;
+  ElementViewConst< arrayView3d< real64 const, compflow::USD_PHASE_DC > > const m_dPhaseVolFrac_dCompDens;
+
+  /// Views on phase mobilities
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_PHASE > > const m_phaseMob;
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_PHASE > > const m_dPhaseMob_dPres;
+  ElementViewConst< arrayView3d< real64 const, compflow::USD_PHASE_DC > > const m_dPhaseMob_dCompDens;
+
+  /// Views on phase mass densities
+  ElementViewConst< arrayView3d< real64 const, multifluid::USD_PHASE > > const m_phaseMassDens;
+  ElementViewConst< arrayView3d< real64 const, multifluid::USD_PHASE > > const m_dPhaseMassDens_dPres;
+  ElementViewConst< arrayView4d< real64 const, multifluid::USD_PHASE_DC > > const m_dPhaseMassDens_dComp;
+
+  /// Views on phase component fractions
+  ElementViewConst< arrayView4d< real64 const, multifluid::USD_PHASE_COMP > > const m_phaseCompFrac;
+  ElementViewConst< arrayView4d< real64 const, multifluid::USD_PHASE_COMP > > const m_dPhaseCompFrac_dPres;
+  ElementViewConst< arrayView5d< real64 const, multifluid::USD_PHASE_COMP_DC > > const m_dPhaseCompFrac_dComp;
+
+  /// Views on phase capillary pressure
+  ElementViewConst< arrayView3d< real64 const, cappres::USD_CAPPRES > > const m_phaseCapPressure;
+  ElementViewConst< arrayView4d< real64 const, cappres::USD_CAPPRES_DS > > const m_dPhaseCapPressure_dPhaseVolFrac;
+
+  // Residual and jacobian
+
+  /// View on the local CRS matrix
+  CRSMatrixView< real64, globalIndex const > const m_localMatrix;
+  /// View on the local RHS
+  arrayView1d< real64 > const m_localRhs;
+};
+
+/**
+ * @class FaceBasedAssemblyKernel
+ * @tparam NUM_COMP number of fluid components
+ * @tparam NUM_DOF number of degrees of freedom
+ * @tparam STENCILWRAPPER the type of the stencil wrapper
+ * @brief Define the interface for the assembly kernel in charge of flux terms
+ */
+template< integer NUM_COMP, integer NUM_DOF, typename STENCILWRAPPER >
+class FaceBasedAssemblyKernel : public FaceBasedAssemblyKernelBase
```

Comment:
Here I extracted a base class `FaceBasedAssemblyKernelBase` (above) which holds all the data independent of the template parameters, including all field accessors. The reason for that is that the constructor is quite large and involves a lot of template instantiations (first, the machinery of `traits::type_list_index<>` and then `std::get<>`) and I'd like to keep it in a separate translation unit (`CompositionalMultiphaseFVMKernels.cpp`) and also avoid it being recompiled for every combination of template parameters here.